web/public/sw.js

```diff
@@ -0,0 +1,180 @@
+          icon: "/static/images/ntfy.png",
+          data,
+        });
+      }
+    })()
+  );
+});
+
+self.addEventListener("notificationclick", (event) => {
+  console.log("[ServiceWorker] NotificationClick");
+
+  event.waitUntil(
+    (async () => {
+      const clients = await self.clients.matchAll({ type: "window" });
+
+      const rootUrl = new URL(self.location.origin);
+      const rootClient = clients.find((client) => client.url === rootUrl.toString());
+
+      if (event.notification.data?.event !== "message") {
+        if (rootClient) {
+          rootClient.focus();
+        } else {
+          self.clients.openWindow(rootUrl);
+        }
+        event.notification.close();
+      } else {
+        const { message } = event.notification.data;
+
+        if (event.action) {
+          const action = event.notification.data.message.actions.find(({ label }) => event.action === label);
+
+          if (action.action === "view") {
+            self.clients.openWindow(action.url);
+
+            if (action.clear) {
+              event.notification.close();
+            }
+          } else if (action.action === "http") {
+            try {
+              const response = await fetch(action.url, {
+                method: action.method ?? "POST",
+                headers: action.headers ?? {},
+                body: action.body,
+              });
+
+              if (!response.ok) {
+                throw new Error(`HTTP ${response.status} ${response.statusText}`);
+              }
+
+              if (action.clear) {
+                event.notification.close();
+              }
+            } catch (e) {
+              console.error("[ServiceWorker] Error performing http action", e);
+              self.registration.showNotification(`Unsuccessful action ${action.label} (${action.action})`, {
+                body: e.message,
+              });
+            }
+          }
+        } else if (message.click) {
+          self.clients.openWindow(message.click);
+
+          event.notification.close();
+        } else {
+          const topicUrl = new URL(message.topic, self.location.origin);
+          const topicClient = clients.find((client) => client.url === topicUrl.toString());
+
+          if (topicClient) {
+            topicClient.focus();
+          } else if (rootClient) {
+            rootClient.focus();
+          } else {
+            self.clients.openWindow(topicUrl);
+          }
+
+          event.notification.close();
+        }
+      }
+    })()
+  );
+});
+
+// self.__WB_MANIFEST is default injection point
+// eslint-disable-next-line no-underscore-dangle
+precacheAndRoute(self.__WB_MANIFEST);
+
+// clean old assets
+cleanupOutdatedCaches();
+
+// to allow work offline
+if (import.meta.env.MODE !== "development") {
+  registerRoute(new NavigationRoute(createHandlerBoundToURL("/")));
+  registerRoute(({ url }) => url.pathname === "/config.js", new NetworkFirst());
+}
```

Comment:
This entire file is a bit messy: Lots of nested, unnamed, async functions, and very little explanation. 

If the service worker supports it, can we 
- break out some functions (could also be inside the `addEventListener`)
- add some docs to the file itself and to the individual functions
?

I can probably do that too, but I think I have to stop working on this for now. I think given that service workers are less common knowledge, it's important to document well what's happening here.

Reply:
Done, let me know what's still unclear or confusing